x/oracle/keeper/grpc_query.go

```diff
@@ -42,93 +42,51 @@ func (q queryServer) GetAllCurrencyPairs(ctx context.Context, _ *types.GetAllCur
 // GetPrice gets the QuotePrice and the nonce for the QuotePrice for a given CurrencyPair. The request contains a
 // CurrencyPairSelector (either the stringified CurrencyPair, or the CurrencyPair itself). If the request is nil this method fails.
 // If the selector is an incorrectly formatted string this method fails. If the QuotePrice / Nonce do not exist for this CurrencyPair, this method fails.
-func (q queryServer) GetPrice(goCtx context.Context, req *types.GetPriceRequest) (_ *types.GetPriceResponse, err error) {
-	var cp types.CurrencyPair
-
+func (q queryServer) GetPrice(goCtx context.Context, req *types.GetPriceRequest) (*types.GetPriceResponse, error) {
 	// fail on nil requests
 	if req == nil {
 		return nil, fmt.Errorf("request cannot be nil")
 	}
-	// determine what type the selector from the response is giving
-	switch cpI := req.CurrencyPairSelector.(type) {
-
-	case *types.GetPriceRequest_CurrencyPairId:
-		// retrieve the currency pair from the stringified ID, and fail if incorrectly formatted
-		cp, err = types.CurrencyPairFromString(cpI.CurrencyPairId)
-		if err != nil {
-			return nil, fmt.Errorf("error unmarshalling CurrencyPairID: %v", err)
-		}
-
-	case *types.GetPriceRequest_CurrencyPair:
-		// retrieve CurrencyPair directly from selector
-		if cpI.CurrencyPair == nil {
-			return nil, fmt.Errorf("currency Pair cannot be nil")
-		}
-		cp = *cpI.CurrencyPair
-
-	default:
-		// fail if any other type of CurrencyPairSelector is given
-		return nil, fmt.Errorf("invalid CurrencyPairSelector given in request (consult documentation)")
-	}
 
 	// unwrap ctx
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// get the QuotePrice + nonce for the given CurrencyPair
-	qpn, err := q.k.GetPriceWithNonceForCurrencyPair(ctx, cp)
+	cps, err := q.k.currencyPairs.Get(ctx, req.CurrencyPairId)
```

Comment:
Since we are only querying by CurrencyPairID, we can simplify the keeper accesses to a single read where we get the `CurrencyPairState`